### PR TITLE
jit: pass majority of spectests.

### PR DESCRIPTION
**Pull request by mathetake**
This commit starts running spectests against our JIT compilation engine 
which we started developing about a month ago.
The engine still does not support `br_if` and `call_indirect` instructions. 
Therefore this commit only enables the tests which doesn't rely on these instructions.

The change includes:

* Enabled memory out of bounds error detection.
* Added protection against call stack exhaustion.
* Correctly handled the edge cases for `i{32,64}.rem_s` instructions.
* Fixed register allocations for the shift instructions.
* Fixed float `eq` and `ne` instructions for NaN values
* Fixed stack unwinding on runtime panics
* plus bunch of small bug fixes

### CODE REVIEW

**Comment by mathetake** [comment]
ok I will repurpose this PR towards spectest against JIT initiation

**Comment by mathetake** [comment]
ok cool every spectest without br_if, call_indirect has now passed!

**Comment by mathetake** [comment]
polish the comments then I will mark this review!

File: wasm/spectests/spec_test.go [hunk 1]
```diff
@@ -16,6 +16,7 @@ import (
 	"github.com/stretchr/testify/require"
```
Comment (mathetake) [inline wasm/spectests/spec_test.go:16]: nothing interesting in this change -- almost all is code move.

File: wasm/jit/jit_amd64.go [hunk 2]
```diff
@@ -1252,19 +1361,78 @@ func (c *amd64Compiler) compileRem(o *wazeroir.OperationRem) (err error) {
...
+	// place the remainder of the result there.
+	if x2.register != remainderRegister {
+		c.onValueReleaseRegisterToStack(remainderRegister)
+	}
+
+	isSignedRem := isRem && signed
+	var signedRemMinusOneDivisorJmp *obj.Prog
+	if isSignedRem {
+		// If this is for getting remainder of signed division,
+		// we have to treat the special case where the divisor equals -1.
```
Comment (mathetake) [inline wasm/jit/jit_amd64.go:1385]: This was one of the hardest parts :D

**Comment by codefromthecrypt** [comment]
image

**Comment by codefromthecrypt** [comment]
by spec tests, you mean tests required for any implementation including web browsers. this is worth noting how comprehensive this is.. this means except the noted TODOs, the JIT is functionally equivalent to the interpreter according to those who control the WebAssembly specification!

File: wasm/jit/jit_amd64.go [hunk 3]
```diff
@@ -3350,90 +3659,181 @@ func (c *amd64Compiler) compileLoad32(o *wazeroir.OperationLoad32) error {
 	return nil
 }
 
+// setupMemoryOffset pops the top value from the stack (called "base"), and returns the result of addition with
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64.go:3662]: good use of param and result names in doc!

File: wasm/jit/jit_amd64_test.go [hunk 4]
```diff
@@ -1350,11 +1370,28 @@ func TestAmd64Compiler_emitEqOrNe(t *testing.T) {
...
+					{x1: math.Inf(1), x2: math.Inf(1)},
+					{x1: math.Inf(-1), x2: 100},
+					{x1: 100, x2: math.Inf(-1)},
+					{x1: math.Inf(-1), x2: math.Inf(-1)},
 				} {
 					t.Run(fmt.Sprintf("%d", i), func(t *testing.T) {
 						compiler := requireNewCompiler(t)
 						compiler.initializeReservedRegisters()
 
+						// For float operations, we have to reserve two int registers to deal with NaN cases.
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64_test.go:1388]: extract to a function to reduce copy/paste drift if possible... I know bit length might not make that work, so just think about it and resolve this comment either way

File: wasm/jit/jit_amd64_test.go [hunk 5]
```diff
@@ -4807,6 +4991,62 @@ func TestAmd64Compiler_compileCall(t *testing.T) {
...
+					eng := newEngine()
+					eng.memroySliceLen = len(mem.Buffer)
+					jitcall(
+						uintptr(unsafe.Pointer(&code[0])),
+						uintptr(unsafe.Pointer(eng)),
+						uintptr(unsafe.Pointer(&mem.Buffer[0])),
+					)
+
+					// If the memory offset exceeds the length of memory, we must exit the function
+					// with jitCallStatusCodeInvalidMemoryOutOfBounds status code.
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64_test.go:5037]: 👍

**Review by codefromthecrypt — APPROVED** [review]
This PR has a humble name, but represents the largest milestone in the project: (almost entirely) working JIT (non-interpreted) mode!

image

**Comment by mathetake** [comment]
@codefromthecrypt thanks for the look! We are almost there 👯

**Comment by codefromthecrypt** [comment]
image
